api/hypershift/v1alpha1/nodepool_types.go

```diff
@@ -911,32 +950,75 @@ type AzureNodePoolPlatform struct {
+
+	// DiskEncryptionSetID is the ID of the DiskEncryptionSet resource to use to encrypt the OS disks for the VMs. This
+	// needs to exist in the same subscription id listed in the Hosted Cluster, hcluster.Spec.Platform.Azure.SubscriptionID.
+	// DiskEncryptionSetID should also exist in a resource group under the same subscription id and the same location
+	// listed in the Hosted Cluster, hcluster.Spec.Platform.Azure.Location.
+	//
 	// +optional
 	DiskEncryptionSetID string `json:"diskEncryptionSetID,omitempty"`
-	// EnableEphemeralOSDisk enables ephemeral OS disk
+
+	// EnableEphemeralOSDisk is a flag when set to true, will enable ephemeral OS disk.
+	//
 	// +optional
 	EnableEphemeralOSDisk bool `json:"enableEphemeralOSDisk,omitempty"`
+
 	// SubnetID is the subnet ID of an existing subnet where the nodes in the nodepool will be created. This can be a
 	// different subnet than the one listed in the HostedCluster, hcluster.Spec.Platform.Azure.SubnetID, but must exist
 	// in the same hcluster.Spec.Platform.Azure.VnetID and must exist under the same subscription ID,
 	// hcluster.Spec.Platform.Azure.SubscriptionID.
 	//
 	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="SubnetID is immutable"
 	// +kubebuilder:validation:Required
-	// +immutable
-	// +required
 	SubnetID string `json:"subnetID"`
+
 	// Diagnostics specifies the diagnostics settings for a virtual machine.
-	// If not specified then Boot diagnostics will be disabled.
+	// If not specified, then Boot diagnostics will be disabled.
 	// +optional
 	Diagnostics *Diagnostics `json:"diagnostics,omitempty"`
 }
 
+// MarketplaceImage specifies the information needed to create an Azure VM from an Azure Marketplace image. This struct
+// replicates the same fields found in CAPZ - https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go.
+type MarketplaceImage struct {
+	// Publisher is the name of the organization that created the image
+	//
+	// +kubebuilder:validation:MinLength=1
+	Publisher string `json:"publisher"`
+
+	// Offer specifies the name of a group of related images created by the publisher.
+	//
+	// +kubebuilder:validation:MinLength=1
+	Offer string `json:"offer"`
+
+	// SKU specifies an instance of an offer, such as a major release of a distribution.
+	// For example, 18.04-LTS, 2019-Datacenter
+	//
+	// +kubebuilder:validation:MinLength=1
+	SKU string `json:"sku"`
+
+	// Version specifies the version of an image sku. The allowed formats are Major.Minor.Build or 'latest'. Major,
+	// Minor, and Build are decimal numbers. Specify 'latest' to use the latest version of an image available at
+	// deployment time. Even if you use 'latest', the VM image will not automatically update after deploy time even if a
+	// new version becomes available.
+	//
+	// +kubebuilder:validation:MinLength=1
+	Version string `json:"version"`
```

Comment:
You should consider for each of these, what a sensible maximum length would be, any associated character sets that are or are not allowed, any patterns that should be matched

Reply:
Ack, I just used the same CEL requirements CAPZ had. 

@bennerv would you happen to know if there are any such patterns, maximum lengths, etc. we should watch out for here?

Reply:
https://learn.microsoft.com/en-us/partner-center/marketplace-offers/product-ingestion-api-vm#resource-api-reference

https://ingestionapi-swagger.azureedge.net/#/
```
$ az vm image list --publisher azureopenshift --offer aro4 --all  -otable
Architecture    Offer    Publisher       Sku      Urn                                          Version
--------------  -------  --------------  -------  -------------------------------------------  ---------------
x64             aro4     azureopenshift  411-v2   azureopenshift:aro4:411-v2:411.86.20220811   411.86.20220811
x64             aro4     azureopenshift  411-v2   azureopenshift:aro4:411-v2:411.86.20221004   411.86.20221004
```

- `publisher` I think this corresponds to externalID in the [product API](https://schema.mp.microsoft.com/schema/product/2022-03-01-preview3)
- "offer" corresponds to the [product API](https://schema.mp.microsoft.com/schema/product/2022-03-01-preview3)
    - This is hard to parse, I can't find it. 
- "sku" corresponds to the [virtual machine plan technical configuration](https://schema.mp.microsoft.com/schema/virtual-machine-plan-technical-configuration/2022-03-01-preview3)
    - pattern: `"^[a-z0-9-_]\u002B$"`
- version corresponds to the [same as above](https://schema.mp.microsoft.com/schema/virtual-machine-plan-technical-configuration/2022-03-01-preview3)
```  "versionNumber": {
          "minLength": 1,
          "maxLength": 32,
          "pattern": "^[0-9]\u002B\\.[0-9]\u002B\\.[0-9]\u002B$"
        },
```

Reply:
Ok so publisher should have (to save ppl clicking a link):
```
            "pattern": "^[a-z0-9][a-z0-9-_]{2,49}$",
              "minLength": 3,
              "maxLength": 50,
```

Reply:
Ok these should be added now. I had to sub out some of the unicode, like `\u002B`